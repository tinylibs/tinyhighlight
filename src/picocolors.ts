import { type TokenColors, highlight as baseHighlight } from './index.js'
import c from 'picocolors'
import type { Colors } from 'picocolors/types.js'

interface ColoredOptions {
  jsx?: boolean
}

function getDefs(c: Colors): TokenColors {
  const Invalid = (text: string) => c.white(c.bgRed(c.bold(text)))
  return {
    Keyword: c.magenta,
    IdentifierCapitalized: c.yellow,
    Punctuator: c.yellow,
    StringLiteral: c.green,
    NoSubstitutionTemplate: c.green,
    MultiLineComment: c.gray,
    SingleLineComment: c.gray,
    RegularExpressionLiteral: c.cyan,
    NumericLiteral: c.blue,
    TemplateHead: (text) => c.green(text.slice(0, text.length - 2)) + c.cyan(text.slice(-2)),
    TemplateTail: (text) => c.cyan(text.slice(0, 1)) + c.green(text.slice(1)),
    TemplateMiddle: (text) => c.cyan(text.slice(0, 1)) + c.green(text.slice(1, text.length - 2)) + c.cyan(text.slice(-2)),
    IdentifierCallable: c.blue,
    PrivateIdentifierCallable: (text) => `#${c.blue(`${text}`.slice(1))}`,
    Invalid,

    JSXString: c.green,
    JSXIdentifier: c.yellow,
    JSXInvalid: Invalid,
    JSXPunctuator: c.yellow,
  }
}

export function highlight(
  code: string,
  options: ColoredOptions = { jsx: false }
) {
  if (code) {
    return baseHighlight(code, {
      jsx: options.jsx,
      colors: getDefs(c),
    })
  }
  return ''
}
