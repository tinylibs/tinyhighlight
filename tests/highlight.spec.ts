import { it, expect } from 'vitest'
import { highlight } from '../src/picocolors.js'

it('should highlight code', () => {
  const code = `
import 'module'
import { test } from 'module'
import * as module from 'module'
import module from 'module'
function test() {}
const test = () => {}
class Test {}
export default test
export const test = 'test'
export function test() {}
export class Test {}
export * as test from 'module'
export * from 'module'
export { test } from 'module'
export { test as test2 } from 'module'
const string = "test"
const string = \`test\`
const string = \`dsdsdt$\{test\}st\`
const string = \`d\$\{SOE\}dsdt$\{test\}st\`
const regexp = /test/
const test = (() => {})()
test?.()
test!()
test ()
test
()
{
  console.log('test')
}
/**
 * test
 */
/** test */
// test
  `
  expect(highlight(code)).toMatchSnapshot()
})

it('should highlight jsx code', () => {
  const code = `
function Test() {
  return <div>test</div>
}

function Test() {
  return <div>{'test'}</div>
}

function Test() {
  return (
    <>
      <div>test</div>
      <div>test</div>
    </>
  )
}

function Invalid() {
  return </div>{test}<div>
}
  `
  expect(highlight(code, { jsx: true })).toMatchSnapshot()
})

it('renders html-like file simillar to jsx', () => {
  const html = `
<div class="test">test</div>
<script setup lang="ts">
console.log("test")
</script>
  `
  expect(highlight(html, { jsx: true })).toMatchSnapshot()
})

it('renders html-like file without jsx setting', () => {
  const html = `
<div class="test">test</div>
<script setup lang="ts">
console.log("test")
</script>
  `
  expect(highlight(html, { jsx: false })).toMatchSnapshot()
})