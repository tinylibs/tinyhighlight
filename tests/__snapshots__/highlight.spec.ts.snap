// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders html-like file simillar to jsx 1`] = `
"
[33m<[39m[33mdiv[39m [33mclass[39m[33m=[39m[32m"test"[39m[33m>[39mtest[33m<[39m[33m/[39m[33mdiv[39m[33m>[39m
[33m<[39mscript setup lang[33m=[39m[32m"ts"[39m[33m>[39m
console[33m.[39m[34mlog[39m([32m"test"[39m)
[33m<[39m[36m/script>[39m
  "
`;

exports[`renders html-like file without jsx setting 1`] = `
"
[33m<[39mdiv [35mclass[39m[33m=[39m[32m"test"[39m[33m>[39mtest[33m<[39m[36m/div>[39m
[33m<[39mscript setup lang[33m=[39m[32m"ts"[39m[33m>[39m
console[33m.[39m[34mlog[39m([32m"test"[39m)
[33m<[39m[36m/script>[39m
  "
`;

exports[`should highlight code 1`] = `
"
[35mimport[39m [32m'module'[39m
[35mimport[39m { test } [35mfrom[39m [32m'module'[39m
[35mimport[39m [33m*[39m [35mas[39m module [35mfrom[39m [32m'module'[39m
[35mimport[39m module [35mfrom[39m [32m'module'[39m
[35mfunction[39m [34mtest[39m() {}
[35mconst[39m test [33m=[39m () [33m=>[39m {}
[35mclass[39m [33mTest[39m {}
[35mexport[39m [35mdefault[39m test
[35mexport[39m [35mconst[39m test [33m=[39m [32m'test'[39m
[35mexport[39m [35mfunction[39m [34mtest[39m() {}
[35mexport[39m [35mclass[39m [33mTest[39m {}
[35mexport[39m [33m*[39m [35mas[39m test [35mfrom[39m [32m'module'[39m
[35mexport[39m [33m*[39m [35mfrom[39m [32m'module'[39m
[35mexport[39m { test } [35mfrom[39m [32m'module'[39m
[35mexport[39m { test [35mas[39m test2 } [35mfrom[39m [32m'module'[39m
[35mconst[39m string [33m=[39m [32m"test"[39m
[35mconst[39m string [33m=[39m [32m\`test\`[39m
[35mconst[39m string [33m=[39m [32m\`dsdsdt[39m[36m\${[39mtest[36m}[39m[32mst\`[39m
[35mconst[39m string [33m=[39m [32m\`d[39m[36m\${[39m[33mSOE[39m[36m}[39m[32mdsdt[39m[36m\${[39mtest[36m}[39m[32mst\`[39m
[35mconst[39m regexp [33m=[39m [36m/test/[39m
[35mconst[39m test [33m=[39m (() [33m=>[39m {})()
[34mtest[39m[33m?.[39m()
[34mtest[39m[33m![39m()
[34mtest[39m ()
[34mtest[39m
()
{
  console[33m.[39m[34mlog[39m([32m'test'[39m)
}
[90m/**
 * test
 */[39m
[90m/** test */[39m
[90m// test[39m
  "
`;

exports[`should highlight jsx code 1`] = `
"
[35mfunction[39m [33mTest[39m() {
  [35mreturn[39m [33m<[39m[33mdiv[39m[33m>[39mtest[33m<[39m[33m/[39m[33mdiv[39m[33m>[39m
}

[35mfunction[39m [33mTest[39m() {
  [35mreturn[39m [33m<[39m[33mdiv[39m[33m>[39m[33m{[39m[32m'test'[39m[33m}[39m[33m<[39m[33m/[39m[33mdiv[39m[33m>[39m
}

[35mfunction[39m [33mTest[39m() {
  [35mreturn[39m (
    [33m<[39m[33m>[39m
      [33m<[39m[33mdiv[39m[33m>[39mtest[33m<[39m[33m/[39m[33mdiv[39m[33m>[39m
      [33m<[39m[33mdiv[39m[33m>[39mtest[33m<[39m[33m/[39m[33mdiv[39m[33m>[39m
    [33m<[39m[33m/[39m[33m>[39m
  )
}

[35mfunction[39m [33mInvalid[39m() {
  [35mreturn[39m [33m<[39m[33m/[39m[33mdiv[39m[33m>[39m{test}[33m<[39m[33mdiv[39m[33m>[39m
[37m[41m[1m}[22m[49m[39m
  "
`;
